zorg/buildbot/process/buildrequest.py

```diff
@@ -22,14 +22,20 @@ def collapseRequests(master, builder, req1, req2):
         ('buildsets', str(req2['buildsetid'])))
 
     # Fetch the buildset properties.
-    selfBuildsetPoperties = yield \
-        master.db.buildsets.getBuildsetProperties(
-            str(req1['buildsetid'])
-        )
-    otherBuildsetPoperties = yield \
-        master.db.buildsets.getBuildsetProperties(
-            str(req2['buildsetid'])
-        )
+    selfBuildsetPoperties = yield master.db.buildsets.getBuildsetProperties(
+        str(req1["buildsetid"])
+    )
+    otherBuildsetPoperties = yield master.db.buildsets.getBuildsetProperties(
+        str(req2["buildsetid"])
+    )
+
+    # If the build is going to be a clean build anyway, we can collapse a clean
+    # build and a non-clean build.
+    if getattr(builder.config.factory, "clean"):
+        if 'clean_obj' in selfBuildsetPoperties:
+            del selfBuildsetPoperties["clean_obj"]
+        if 'clean_obj' in otherBuildsetPoperties:
+            del otherBuildsetPoperties["clean_obj"]
```

Comment:
Does `master.db.buildsets.getBuildsetProperties` yields a buildset copy?

If not, wouldn't it be better to instead set `clean_obj` to `True` in both buildsets, just in case other part of the code relies on it?

Reply:
Makes me wonder whether rest of the code depends on factory attribute "clean" or the build set properties clean_obj. From this change it looks like clean_obj was taking precedence over factory attribute hence the delete option was used.

Also I am thinking whether we should somehow restrict the number of builds that can be collapsed together. If queue is long may be only collapse n builds. Something that we can do in a later patch if anything drastic comes up during production builds.

Reply:
> Also I am thinking whether we should somehow restrict the number of builds that can be collapsed together. If queue is long may be only collapse n builds. Something that we can do in a later patch if anything drastic comes up during production builds.

I suspect this requires changes to Buildbot itself.